src/charm.py

```diff
@@ -0,0 +1,595 @@
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.reconcile_runners, self._on_reconcile_runners)
+        self.framework.observe(self.on.update_runner_bin, self._on_update_runner_bin)
+        self.framework.observe(self.on.stop, self._on_stop)
+
+        self.framework.observe(self.on.check_runners_action, self._on_check_runners_action)
+        self.framework.observe(self.on.reconcile_runners_action, self._on_reconcile_runners_action)
+        self.framework.observe(self.on.flush_runners_action, self._on_flush_runners_action)
+        self.framework.observe(self.on.update_runner_bin_action, self._on_update_runner_bin)
+
+    def _get_runner_manager(
+        self, token: Optional[str] = None, path: Optional[str] = None
+    ) -> Optional[RunnerManager]:
+        """Get a RunnerManager instance, or None if missing config.
+
+        Args:
+            token: GitHub personal access token to manager the runners with.
+            path: GitHub repository path in the format '<org>/<repo>', or the GitHub organization
+                name.
+
+        Returns:
+            A instance of RunnerManager if the token and path configuration can be found.
+        """
+        if token is None:
+            token = self.config["token"]
+        if path is None:
+            path = self.config["path"]
+
+        if not token or not path:
+            return None
+
+        if self.service_token is None:
+            self.service_token = self._get_service_token()
+
+        if "/" in path:
+            paths = path.split("/")
+            if len(paths) != 2:
+                logger.error("Invalid path %s", path)
+                return None
+
+            owner, repo = paths
+            path = GitHubRepo(owner=owner, repo=repo)
+        else:
+            path = GitHubOrg(org=path, group=self.config["group"])
+
+        app_name, unit = self.unit.name.rsplit("/", 1)
+        return RunnerManager(
+            app_name,
+            unit,
+            RunnerManagerConfig(path, token, "jammy", self.service_token),
+            proxies=self.proxies,
+        )
+
+    @catch_unexpected_charm_errors
+    def _on_install(self, _event: InstallEvent) -> None:
+        """Handle the installation of charm.
+
+        Args:
+            event: Event of installing charm.
+        """
+        self.unit.status = MaintenanceStatus("Installing packages")
+
+        try:
+            # The `_start_services`, `_install_deps` includes retry.
+            self._install_deps()
+            self._start_services()
+        except SubprocessError as err:
+            logger.exception(err)
+            # The charm cannot proceed without dependencies.
+            self.unit.status = BlockedStatus("Failed to install dependencies")
+            return
+
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
```

Comment:
Minor nitpick, I think it would be better if the not case can come earlier so that it can set BlockedStatus and return first to reduce indentation.
i.e.
```suggestion
if not runner_manager:
    self.unit.status = BlockedStatus(...)
    return
... # continue with execution flow
```

Reply:
Minor comments